### PR TITLE
jstree._append_json_data() callback always asynchronous, even when no worker is used

### PR DESCRIPTION
**Pull request by duzun**
This PR is a fix for #2258 issue, according to my tests on Google Chrome.
It basically uses `Promise` in all major browsers and `setImmediate` where available, to make `jstree._append_json_data()` always async.

### CODE REVIEW

**Comment by duzun** [comment]
This has the added benefit of making UI more responsive while processing a lot of nodes, and possibly a better option than using a worker.

**Comment by duzun** [comment]
I propose to use `setImmediate(cb)` instead of `setTimeout(cb, 0)` in places where wait time is not desired.
Currently `setTimeout(cb, 0)` deafults to 4ms wait after 5th nested call.

**Comment by vakata** [comment]
Thank you for this PR - looks great and seems fully backward compatible. It really is a nice alternative to the worker (which has its own problems). I do have one question though - how much slower is the new version on average? Freeing the UI thread is great, and I do prefer the async approach in all my projects, but there are some use cases, where devs prefer to block the UI, but deliver the tree faster. I need that information before I merge - if you have not tested - I will test it myself and merge.

**Comment by duzun** [comment]
I did not do an extensive test. 
I've just tested in the context of my app and measured the overall time of a search, since start till search callback invoked.
On my dataset, with ~1400 nodes to open and ~4000 nodes to load, it takes ~4 sec with or without this PR. With worker enabled it is ~2 min.
The UI isn't smooth but isn't completely blocked either during the 4 sec.
I also use [setImmediate](https://github.com/YuzuJS/setImmediate) polyfill.
With `setImmediate`/`Promise`, the delay between invocation and execution is one cycle, which is in the range of nano- or microseconds (not 100% sure), when there are no other tasks on the queue (like UI, I/O, GC), and with this PR there are few invocations to `setImmediate` (~1400 in my case I guess), which would not add up to even some 10ms.

So, this doesn't free the UI thread much but at least solves the `Maximum call stack size exceeded error` completely and gives the UI and other tasks some chance of execution.

Maybe using this technique in other places where there is a synchronous callback invocation in a chain would improve the responsiveness of the UI even more.

If I were to use `setTiemout(cb, 0)`, I suppose it would take considerably more time, because of the 4ms delay, but would also make the UI much more responsive. I've used this technique in one of my projects successfully.

So, basically, this PR is a solution to #2258 and a proof of concept on which we could extend and improve the jsTree further, I hope.

**Comment by vakata** [comment]
That is great - thank you for the information. We should probably adopt this approach for v.4 as well.
